--- a/drivers/clk/Makefile
+++ b/drivers/clk/Makefile
@@ -124,6 +124,7 @@ obj-$(CONFIG_COMMON_CLK_PXA)		+= pxa/
 obj-$(CONFIG_COMMON_CLK_QCOM)		+= qcom/
 obj-y					+= ralink/
 obj-$(CONFIG_COMMON_CLK_REALTEK)	+= realtek/
+obj-$(CONFIG_COMMON_CLK_RTK_APRO)	+= realtek/
 obj-y					+= renesas/
 obj-$(CONFIG_ARCH_ROCKCHIP)		+= rockchip/
 obj-$(CONFIG_COMMON_CLK_SAMSUNG)	+= samsung/
--- a/drivers/clk/realtek/Kconfig
+++ b/drivers/clk/realtek/Kconfig
@@ -17,3 +17,9 @@ config COMMON_CLK_RTL83XX
 	  RT8391, RTL8392, RTL8393 and RTL8396.
 
 endif
+
+config COMMON_CLK_RTK_APRO
+	bool "RTK ApolloPro Clock Driver"
+	depends on CLKSRC_MIPS_GIC
+	help
+	  Build the RTK ApolloPro clock driver based on the common clock framework.
--- a/drivers/clk/realtek/Makefile
+++ b/drivers/clk/realtek/Makefile
@@ -1,2 +1,3 @@
 # SPDX-License-Identifier: GPL-2.0-only
 obj-$(CONFIG_COMMON_CLK_RTL83XX) += clk-rtl83xx.o clk-rtl838x-sram.o clk-rtl839x-sram.o
+obj-$(CONFIG_COMMON_CLK_RTK_APRO) += rtk-clk-apro-ocp.o
