--- a/arch/mips/Kbuild.platforms
+++ b/arch/mips/Kbuild.platforms
@@ -21,6 +21,7 @@ platform-$(CONFIG_MIPS_MALTA)		+= mti-ma
 platform-$(CONFIG_MACH_NINTENDO64)	+= n64/
 platform-$(CONFIG_PIC32MZDA)		+= pic32/
 platform-$(CONFIG_RALINK)		+= ralink/
+platform-$(CONFIG_RTL_MIPS_SOC)     += rtl9607c/
 platform-$(CONFIG_MIKROTIK_RB532)	+= rb532/
 platform-$(CONFIG_MACH_REALTEK_RTL)	+= rtl838x/
 platform-$(CONFIG_SGI_IP22)		+= sgi-ip22/
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -670,6 +670,30 @@ config MACH_REALTEK_RTL
 	select ARCH_HAS_RESET_CONTROLLER
 	select RESET_CONTROLLER
 
+config RTL_MIPS_SOC
+	bool "Realtek MIPS series SoCs"
+	select ARCH_HAS_PHYS_TO_DMA
+	select BOOT_ELF32
+	select COMMON_CLK
+	select CLKSRC_MIPS_GIC
+	select DMA_NONCOHERENT
+	select HAVE_PCI
+	select IRQ_MIPS_CPU
+	select MIPS_CPU_SCACHE
+	select MIPS_GIC
+	select SYS_HAS_CPU_MIPS32_R2
+	select SYS_SUPPORTS_BIG_ENDIAN
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_HAS_EARLY_PRINTK
+	select SYS_SUPPORTS_ARBIT_HZ
+	select SYS_SUPPORTS_HIGHMEM
+	select SYS_SUPPORTS_SMP
+	select SYS_SUPPORTS_MULTITHREADING
+	select SYS_SUPPORTS_MIPS_CPS
+	select SYS_SUPPORTS_ZBOOT
+	select USE_OF
+	select WEAK_REORDERING_BEYOND_LLSC
+
 config SGI_IP22
 	bool "SGI IP22 (Indy/Indigo2)"
 	select ARC_MEMORY
@@ -1042,6 +1066,17 @@ config RTL931X
 	select SYS_SUPPORTS_SMP
 	select SYS_SUPPORTS_MIPS_CPS
 
+config RTL9607C
+	bool "Realtek RTL9607C based platforms"
+	depends on RTL_MIPS_SOC
+	select CPU_SUPPORTS_32BIT_KERNEL
+	select CPU_HAS_WB
+	select COMMON_CLK_RTK_APRO
+	select CPU_MIPSR2_IRQ_VI
+	select MIPS_DISABLE_VDSO
+	select SYS_HAS_EARLY_PRINTK_8250
+	select USE_GENERIC_EARLY_PRINTK_8250
+
 source "arch/mips/alchemy/Kconfig"
 source "arch/mips/ath25/Kconfig"
 source "arch/mips/ath79/Kconfig"
@@ -1055,6 +1090,7 @@ source "arch/mips/lantiq/Kconfig"
 source "arch/mips/mobileye/Kconfig"
 source "arch/mips/pic32/Kconfig"
 source "arch/mips/ralink/Kconfig"
+source "arch/mips/rtl9607c/Kconfig"
 source "arch/mips/sgi-ip27/Kconfig"
 source "arch/mips/sibyte/Kconfig"
 source "arch/mips/txx9/Kconfig"
--- a/arch/mips/include/asm/thread_info.h
+++ b/arch/mips/include/asm/thread_info.h
@@ -76,6 +76,9 @@ register unsigned long current_stack_poi
 #endif /* !__ASSEMBLY__ */
 
 /* thread information allocation */
+#ifdef CONFIG_KERNEL_STACK_SIZE_ORDER
+#define THREAD_SIZE_ORDER (CONFIG_KERNEL_STACK_SIZE_ORDER)
+#else
 #if defined(CONFIG_PAGE_SIZE_4KB) && defined(CONFIG_32BIT)
 #define THREAD_SIZE_ORDER (1)
 #endif
@@ -94,6 +97,7 @@ register unsigned long current_stack_poi
 #ifdef CONFIG_PAGE_SIZE_64KB
 #define THREAD_SIZE_ORDER (0)
 #endif
+#endif /* CONFIG_KERNEL_STACK_SIZE_ORDER */
 
 #define THREAD_SIZE (PAGE_SIZE << THREAD_SIZE_ORDER)
 #define THREAD_MASK (THREAD_SIZE - 1UL)
--- a/arch/mips/kernel/vmlinux.lds.S
+++ b/arch/mips/kernel/vmlinux.lds.S
@@ -86,6 +86,20 @@ SECTIONS
 	_sdata = .;			/* Start of data section */
 	RO_DATA(4096)
 
+	. = ALIGN(4);
+	.nand_spif_probe_func : {
+		LS_start_of_nand_spif_probe_func = .;
+		*(.nand_spif_probe_func);
+		LS_end_of_nand_spif_probe_func = .;
+	}
+
+	. = ALIGN(4);
+	.snaf_init_rest_func : {
+		LS_start_of_snaf_init_rest_func = .;
+		KEEP(*(.snaf_init_rest_func));
+		LS_end_of_snaf_init_rest_func = .;
+	}
+
 	/* writeable */
 	.data : {	/* Data */
 		. = . + DATAOFFSET;		/* for CONFIG_MAPPED_KERNEL */
